# Brassica oleracea/rapa/napus code
Code for the Brassica oleracea/rapa/napus genomic comparison.

Most of these folders are RStudio projects that can be opened in RStudio.

## Assembly Annotation

This folder collects small scripts and documentation as to how I ran the assembly and annotation steps for the pangenome.

## GO enrichment

Code and RStudio project is under `GO_Enrichment/`

## Resistance gene analysis

R-code for the stacked barplots for the R-genes can be found in `R_Gene_Analysis/`

## Pangenome modeling

R-code for pangenome size modeling is in the folder `Modeling/`

## Venn diagram plots

R-code for the Venn-diagram plots are in `FinalPanPanVenn/`

## Core vs. variable genes plots

R-code for the variable genes vs core genes is in `Visualise_Core_Variable_A_vs_C/`

## PCA plots

R-code for the PCA plots and animations are in `Cluster_Napus_A_and_C_PAV/`

## Notebooks 

These contain the XGBoost and Shapley values work used for modeling gene presence/absence variation.

You can launch them on binder:
[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/appliedbioinformatics/Brassica_oleracea_rapa_napus_code/master?filepath=notebooks)

Or navigate to the `notebooks` folder.
