# First, setup

```{r setup}
library(tidyverse)
library(cowplot)
library(minpack.lm)
library(broom)
library(rstan)
library(car)
library(patchwork)


options('scipen'=100)
```


## Oleracea model

Here I load the oleracea data and plot it for oleracea only
```{r oleracea_model}
oleracea_df <- read_tsv('Oleracea_Permutations_Table.csv')
head(oleracea_df)

# approximate the para0p
lm(logit(pangenome/76000) ~ Combinations, oleracea_df)

para0p <- c(A=76000,B=-4.6,C=0.0005)
```
OK we got the initaliser

```{r oleracea_training}


oleracea_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=oleracea_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))

summary(oleracea_summary_fitp)
tidy(oleracea_summary_fitp)
```
Now come the core genes!

```{r oleracea_core}
para0c <- c(B=1)
oleracea_fitc <- nls(core~cbind(1, exp(B*Combinations)),oleracea_df, start=para0c,             trace=F, alg='plinear')
tidy(oleracea_fitc)

```

```{r oleracea_plt}

oleracea_fit_data_p <- augment(oleracea_summary_fitp)
oleracea_fit_data_c <- augment(oleracea_fitc)

oleracea_p <-ggplot() + 
  geom_point(data=oleracea_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=oleracea_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=oleracea_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) + 
  geom_line(data=oleracea_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) + 
  #xlim(c(0,50)) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio = 1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")
oleracea_p
```

## Rapa model

```{r rapa_model}
rapa_df <- read_tsv('Rapa_Permutations_Table.csv')
head(rapa_df)

# approximate the para0p

lm(logit(pangenome/63000) ~ Combinations, rapa_df)

para0p <- c(A=62000,B=-4.6,C=0.0006)
```
OK we got the initaliser

```{r rapa_training}


rapa_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=rapa_df, 
                      start=para0p, trace=T, control=list(maxiter=200, maxfev=1500))

summary(rapa_summary_fitp)
tidy(rapa_summary_fitp)
```
Now come the core genes!

```{r rapa_core}
head(rapa_df)
para0c <- c(B=1)
rapa_fitc <- nls(core~cbind(1, exp(B*Combinations)),rapa_df, start=para0c, 
            trace=F, alg='plinear')
tidy(rapa_fitc)
```

```{r rapa_plot}
rapa_fit_data_p <- augment(rapa_summary_fitp)
rapa_fit_data_c <- augment(rapa_fitc)

rapa_p <- ggplot() + 
  geom_point(data=rapa_df, aes(x=Combinations, y=pangenome), alpha=0.1) +
  geom_point(data=rapa_df, aes(x=Combinations, y=core), alpha=0.1) +
  geom_line(data=rapa_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=rapa_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio = 1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")
rapa_p
```



## Napus model

```{r napus_model}
#napus_df <- read_tsv('Napus_Permutations_Table.csv')

napus_df <- read_tsv('Napus_Permutations_Table_with_72_row.csv')
head(napus_df)

# approximate the para0p

lm(logit(pangenome/max(napus_df$pangenome)) ~ Combinations, napus_df)

para0p <- c(A=max(napus_df$pangenome),B=-3.367726,C=0.005872)

```
OK we got the initaliser

```{r napus_training}

napus_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=napus_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))

summary(napus_summary_fitp)
tidy(napus_summary_fitp)
```

Now come the core genes!

```{r napus_core}

para0c <- c(B=1)
napus_fitc <- nls(core~cbind(1, exp(B*Combinations)),napus_df, start=para0c, 
            trace=F, alg='plinear')
tidy(napus_fitc)

```


```{r napus_plot}
napus_fit_data_p <- augment(napus_summary_fitp)
napus_fit_data_c <- augment(napus_fitc)

napus_p <- ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=napus_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=napus_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")

napus_p
```

# Patchworking!

```{r patches}
grid <- plot_grid(oleracea_p, rapa_p, napus_p, labels='AUTO')
grid
save_plot(filename= 'Grid.png', plot = grid, base_height = 8)
```

# All in one!

```{r all_in_one}
ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_point(data=rapa_df, aes(x=Combinations, y=pangenome), alpha=0.1) +
    geom_line(data=rapa_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_point(data=oleracea_df, aes(x=Combinations, y=pangenome), alpha=0.1) +
    geom_line(data=oleracea_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none") +
  xlim(c(0, 50))

```

yeah let's not use that :P



# Another run - this time with different data? Maybe?
## Oleracea 2nd model

Here I load the oleracea data and plot it for oleracea only
```{r oleracea_model}
oleracea_df <- read_tsv('./Oleracea_86_table.txt')
head(oleracea_df)

# approximate the para0p
lm(logit(pangenome/max(oleracea_df$pangenome)) ~ Combinations, oleracea_df)

para0p <- c(A=max(oleracea_df$pangenome)+1000,B=-3.47,C=0.003)
```
OK we got the initaliser

```{r oleracea_training}


oleracea_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=oleracea_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))

summary(oleracea_summary_fitp)
tidy(oleracea_summary_fitp)
```
Now come the core genes!

```{r oleracea_core}
para0c <- c(B=1)
oleracea_fitc <- nls(core~cbind(1, exp(B*Combinations)),oleracea_df, start=para0c, trace=F, alg='plinear')
tidy(oleracea_fitc)

```

Can we add in fake values for really large x??
```{r}
new_combinations <- data.frame(Combinations = 1:1000, type.predict='response')

tail(augment(oleracea_summary_fitp, newdata=new_combinations))
```
That doesn't help much :)

```{r oleracea_plt}

oleracea_fit_data_p <- augment(oleracea_summary_fitp)
oleracea_fit_data_c <- augment(oleracea_fitc)

oleracea_p <-ggplot() + 
  geom_point(data=oleracea_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=oleracea_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=oleracea_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) + 
  geom_line(data=oleracea_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) + 
  #xlim(c(0,50)) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio = 1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")
oleracea_p
```
## Rapa 2nd model

```{r rapa_model}
rapa_df <- read_tsv('./Rapa_71_table.txt')
head(rapa_df)

# approximate the para0p

lm(logit(pangenome/max(rapa_df$pangenome)) ~ Combinations, rapa_df)

para0p <- c(A=max(rapa_df$pangenome)+1000,B=-3.3,C=0.006)
```
OK we got the initaliser

```{r rapa_training}
sum_rapa_df <- rapa_df %>% group_by(Combinations) %>% summarise(mean_pangenome = mean(pangenome))
sum_rapa_df


rapa_summary_fitp <- nlsLM(mean_pangenome~A*Combinations^B+C, data=sum_rapa_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
#rapa_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=rapa_df, 
                      #start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))

summary(rapa_summary_fitp)
tidy(rapa_summary_fitp)
```
Now come the core genes!

```{r rapa_core}
head(rapa_df)
para0c <- c(B=1)
rapa_fitc <- nls(core~cbind(1, exp(B*Combinations)),rapa_df, start=para0c, 
            trace=F, alg='plinear')
tidy(rapa_fitc)
```

```{r rapa_plot}
rapa_fit_data_p <- augment(rapa_summary_fitp)
rapa_fit_data_c <- augment(rapa_fitc)

rapa_p <- ggplot() + 
  geom_point(data=rapa_df, aes(x=Combinations, y=pangenome), alpha=0.1) +
  geom_point(data=rapa_df, aes(x=Combinations, y=core), alpha=0.1) +
  geom_line(data=rapa_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=rapa_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio = 1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")
rapa_p
```

## Napus 2nd model

```{r napus_model}
napus_df <- read_tsv('./Napus_72_table.txt')
tail(napus_df)

# approximate the para0p

lm(logit(pangenome/132862) ~ Combinations, napus_df)



```
OK we got the initaliser

```{r napus_training}
para0p <- c(A=132862, B=-3.5,C=0.003)
para0p <- c(A=132862, B=-2.5,C=0.003)

napus_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=napus_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
summary(napus_summary_fitp)
tidy(napus_summary_fitp)
```

Now come the core genes!

```{r napus_core}

para0c <- c(B=1)
napus_fitc <- nls(core~cbind(1, exp(B*Combinations)),napus_df, start=para0c, 
            trace=F, alg='plinear')
tidy(napus_fitc)

```


```{r napus_plot}
napus_fit_data_p <- augment(napus_summary_fitp)
napus_fit_data_c <- augment(napus_fitc)

napus_p <- ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=napus_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=napus_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")

napus_p
```

# Patchworking!

```{r patches}
top_row <- plot_grid(oleracea_p, rapa_p, ncol=2, labels=c('A','B'))
bottom_row <- plot_grid(NULL, napus_p, NULL,ncol=3,
                        rel_widths=c(0.25,0.5,0.25),
                        label_x = 1, labels='C')
grid <- plot_grid(top_row, bottom_row, ncol=1)
save_plot(filename= 'Grid.png', plot = grid, base_height = 8)
```




## Napus no synths model

```{r napus_model}
napus_df <- read_tsv('./Napus_NoSynths_54_table.txt')
tail(napus_df)

# approximate the para0p

lm(logit(pangenome/132719) ~ Combinations, napus_df)



```
OK we got the initaliser

```{r napus_training}
para0p <- c(A=132719, B=-3.411,C=0.0061)


napus_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=napus_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
summary(napus_summary_fitp)
tidy(napus_summary_fitp)
```

Now come the core genes!

```{r napus_core}

para0c <- c(B=1)
napus_fitc <- nls(core~cbind(1, exp(B*Combinations)),napus_df, start=para0c, 
            trace=F, alg='plinear')
tidy(napus_fitc)

```


```{r napus_plot}
napus_fit_data_p <- augment(napus_summary_fitp)
napus_fit_data_c <- augment(napus_fitc)

napus_nonsyn_p <- ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=napus_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=napus_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")

napus_nonsyn_p
```


## Patchworking with four plots

```{r patches}
top_row <- plot_grid(oleracea_p, rapa_p, ncol=2, labels=c('A','B'))
bottom_row <- plot_grid(napus_p, napus_nonsyn_p, ncol=2,
                        
                        labels=c('C', 'D'))
grid <- plot_grid(top_row, bottom_row, ncol=1)
grid
save_plot(filename= 'Grid_synths_nonsynths.png', plot = grid, base_height = 8)
```


# REDUCED tree model


## REDUCED Napus

```{r napus_model}
napus_df <- read_tsv('./../../PAV_Phylogeny/Napus_72_20_subset.txt')
tail(napus_df)

# approximate the para0p

lm(logit(pangenome/max(napus_df$pangenome)) ~ Combinations, napus_df)


```
OK we got the initaliser

```{r napus_training}
para0p <- c(A=max(napus_df$pangenome)-5000, B=-3.36982,C=0.02786)


napus_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=napus_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
summary(napus_summary_fitp)
tidy(napus_summary_fitp)
```

Now come the core genes!

```{r napus_core}

para0c <- c(B=1)
napus_fitc <- nls(core~cbind(1, exp(B*Combinations)),napus_df, start=para0c, 
            trace=F, alg='plinear')
tidy(napus_fitc)

```


```{r napus_plot}
napus_fit_data_p <- augment(napus_summary_fitp)
napus_fit_data_c <- augment(napus_fitc)

napus_nonsyn_p <- ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=napus_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=napus_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")

napus_nonsyn_p
```

## REDUCED Oleracea

```{r oleracea_model}
oleracea_df <- read_tsv('./../../PAV_Phylogeny/Oleracea_86_20_subset_with_ALL.txt')


# approximate the para0p

lm(logit(pangenome/max(napus_df$pangenome)) ~ Combinations, oleracea_df)


```
OK we got the initaliser

```{r oleracea_training}


summary_oleracea_df <- oleracea_df %>% group_by(Combinations) %>% summarise(mean_pangenome = mean(pangenome))
summary_oleracea_df
lm(logit(mean_pangenome/max(summary_oleracea_df$mean_pangenome)) ~ Combinations, summary_oleracea_df)
para0p <- c(A=max(summary_oleracea_df$mean_pangenome)-1000, B=-3.316,C=0.004624)

oleracea_summary_fitp <- nlsLM(mean_pangenome~A*Combinations^B+C, data=summary_oleracea_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
summary(oleracea_summary_fitp)
tidy(oleracea_summary_fitp)
```

# FINAL - with manually added last line with 'all' core and variable genes


## Napus model

```{r napus_model}
napus_df <- read_tsv('./../../Core_Variable_Gene_Analysis/Modeling/Napus_72_table_with_72_row_table.txt')
tail(napus_df)

# approximate the para0p

lm(logit(pangenome/max(napus_df$pangenome)) ~ Combinations, napus_df)


```
OK we got the initaliser

```{r napus_training}
para0p <- c(A=max(napus_df$pangenome)+1000, B=-3.492556,C=0.003025)


napus_summary_fitp <- nlsLM(pangenome~A*Combinations^B+C, data=napus_df, 
                      start=para0p, trace=F, control=list(maxiter=200, maxfev=1500))
summary(napus_summary_fitp)
tidy(napus_summary_fitp)
```

Now come the core genes!

```{r napus_core}

para0c <- c(B=1)
napus_fitc <- nls(core~cbind(1, exp(B*Combinations)),napus_df, start=para0c, 
            trace=F, alg='plinear')
tidy(napus_fitc)

```


```{r napus_plot}
napus_fit_data_p <- augment(napus_summary_fitp)
napus_fit_data_c <- augment(napus_fitc)

napus_nonsyn_p <- ggplot() + 
  geom_point(data=napus_df, aes(x=Combinations, y=pangenome), alpha=0.1) + 
  geom_point(data=napus_df, aes(x=Combinations, y=core), alpha=0.1) + 
  geom_line(data=napus_fit_data_p, aes(x=Combinations, y=.fitted, colour='red')) +
  geom_line(data=napus_fit_data_c, aes(x=Combinations, y=.fitted, colour='red')) +
  ylab('Number of pangenome genes') +
  xlab('Combinations of individuals') +
  theme(legend.position = "none") +
  theme(aspect.ratio=1) +
  theme_half_open() +
  theme_minimal_hgrid() +
  theme(legend.position="none")

napus_nonsyn_p
```
